src/Ocelot/Request/Mapper/StreamHttpContent.cs

```diff
@@ -0,0 +1,118 @@
+﻿using Microsoft.AspNetCore.Http;
+using System.Buffers;
+
+namespace Ocelot.Request.Mapper;
+
+public class StreamHttpContent : HttpContent
+{
+    private const int DefaultBufferSize = 65536;
+    public const long UnknownLength = -1;
+    private readonly HttpContext _context;
+
+    public StreamHttpContent(HttpContext context)
+    {
+        _context = context ?? throw new ArgumentNullException(nameof(context));
+    }
+
+    protected override async Task SerializeToStreamAsync(Stream stream, TransportContext context,
+        CancellationToken cancellationToken)
+    {
+        await CopyAsync(_context.Request.Body, stream, Headers.ContentLength ?? UnknownLength, false,
+            cancellationToken);
+    }
+
+    protected override async Task SerializeToStreamAsync(Stream stream, TransportContext context)
+    {
+        await CopyAsync(_context.Request.Body, stream, Headers.ContentLength ?? UnknownLength, false,
+            CancellationToken.None);
+    }
+
+    protected override bool TryComputeLength(out long length)
+    {
+        length = -1;
+        return false;
+    }
+
+    // this is used internally by HttpContent.ReadAsStreamAsync(...)
+    protected override Task<Stream> CreateContentReadStreamAsync()
+    {
+        // Nobody should be calling this...
+        throw new NotImplementedException();
+    }
+
+    private static async Task CopyAsync(Stream input, Stream output, long announcedContentLength,
+        bool autoFlush, CancellationToken cancellation)
+    {
+        // for smaller payloads, avoid allocating a buffer that is larger than the announced content length
+        var minBufferSize = announcedContentLength != UnknownLength && announcedContentLength < DefaultBufferSize
+            ? (int)announcedContentLength
+            : DefaultBufferSize;
+
+        var buffer = ArrayPool<byte>.Shared.Rent(minBufferSize);
```

Comment:
Very significant conclusion! Because the logic says "Larger buffer should show better performance".
But in reality it doesn't work for large streams... and streaming scenarios.
🆗 In my opinion, this is a challenge to guess the best buffer size. We can only define very common buffer size...
Some routes can proxy small bodies, some routes can pump large content in streaming scenario...
I would say, it is a bit wrong to define one common buffer size for all routes of Ocelot app instance...
Did you get my thought?

Reply:
Yes, it's the rule of thumb here...